File: tailscale.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

type tailscaleDevice struct {
	Hostname  string   `json:"hostname"`
	Addresses []string `json:"addresses"`
}

type tailscaleDevicesResponse struct {
	Devices []tailscaleDevice `json:"devices"`
}

func fetchTailscaleDevices(apiKey, tailnet string) (map[string]string, error) {
	u := fmt.Sprintf("https://api.tailscale.com/api/v2/tailnet/%s/devices", url.PathEscape(tailnet))
	req, err := http.NewRequest(http.MethodGet, u, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("tailscale API error: %d %s", resp.StatusCode, body)
	}

	var data tailscaleDevicesResponse
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, err
	}

	devices := make(map[string]string)
	for _, d := range data.Devices {
		for _, addr := range d.Addresses {
			if strings.HasPrefix(addr, "100.") {
				devices[strings.ToLower(d.Hostname)] = addr
				break
			}
		}
	}
	return devices, nil
}

File: src/tailscale.ts
interface TailscaleDevice {
  hostname: string
  addresses: string[]
}

interface TailscaleDevicesResponse {
  devices: TailscaleDevice[]
}

export async function fetchTailscaleDevices(apiKey: string, tailnet: string): Promise<Map<string, string>> {
  const url = `https://api.tailscale.com/api/v2/tailnet/${encodeURIComponent(tailnet)}/devices`
  const res = await fetch(url, {
    headers: {
      Authorization: `Bearer ${apiKey}`,
    },
  })
  if (!res.ok) {
    throw new Error(`Tailscale API error: ${res.status} ${await res.text()}`)
  }
  const data = (await res.json()) as TailscaleDevicesResponse
  const deviceMap = new Map<string, string>()
  for (const device of data.devices) {
    const tsIP = device.addresses.find((a) => a.startsWith('100.'))
    if (!tsIP) continue
    const name = device.hostname.toLowerCase()
    deviceMap.set(name, tsIP)
  }
  return deviceMap
}
